third_party/grpc-rust/grpc-compiler/src/codegen.rs

```diff
@@ -0,0 +1,360 @@
+            .message
+            == protobuf::well_known_types::Empty::descriptor_static().get_proto()
+    }
+
+    fn output_empty(&self) -> bool {
+        self.root_scope
+            .find_message(self.proto.get_output_type())
+            .message
+            == protobuf::well_known_types::Empty::descriptor_static().get_proto()
+    }
+
+    fn input_message(&self) -> String {
+        let msg = self.root_scope.find_message(self.proto.get_input_type());
+        let empty = protobuf::well_known_types::Empty::descriptor_static();
+        if msg.message == empty.get_proto() {
+            "()".to_string()
+        } else {
+            format!("super::{}", msg.rust_fq_name())
+        }
+    }
+
+    fn output_message(&self) -> String {
+        let msg = self.root_scope.find_message(self.proto.get_output_type());
+        let empty = protobuf::well_known_types::Empty::descriptor_static();
+        if msg.message == empty.get_proto() {
+            "()".to_string()
+        } else {
+            format!("super::{}", msg.rust_fq_name())
+        }
+    }
+
+    fn server_req_type(&self) -> String {
+        match self.proto.get_client_streaming() {
+            false => format!("{}", self.input_message()),
+            // TODO: better streaming
+            true => format!("Vec<{}>", self.input_message()),
+        }
+    }
+
+    fn server_resp_type(&self) -> String {
+        match self.proto.get_server_streaming() {
+            false => format!("{}", self.output_message()),
+            // TODO: better streaming
+            true => format!("Vec<{}>", self.output_message()),
+        }
+    }
+
+    fn server_sig(&self) -> String {
+        let arg;
+        if self.input_empty() {
+            arg = "".to_string();
+        } else {
+            arg = format!(
+                ", {}: {}",
+                if self.proto.get_client_streaming() {
+                    "reqs"
+                } else {
+                    "req"
+                },
+                self.server_req_type(),
+            );
+        }
+        let result;
+        if self.output_empty() {
+            result = "".to_string();
+        } else {
+            result = format!(" -> {}", self.server_resp_type());
+        }
+        format!("{}(&mut self{}){}", self.snake_name(), arg, result)
+    }
+
+    fn write_server_intf(&self, w: &mut CodeWriter) {
+        w.fn_def(&self.server_sig())
+    }
+
+    fn write_dispatch(&self, w: &mut CodeWriter) {
+        let param_in;
+        if self.input_empty() {
+            param_in = "";
+        } else {
+            match self.proto.get_client_streaming() {
+                false => {
+                    param_in = "req";
+                    w.write_line("let req = protobuf::parse_from_reader(grpc_channel).unwrap();")
+                }
+                true => {
+                    param_in = "reqs";
+                    w.write_line("let mut reqs = vec![];");
```

Comment:
I wonder if we could use the [`quote`](https://github.com/dtolnay/quote) crate for this.

Reply:
`quote` seems nicer, but wouldn't that basically involve starting from scratch?

Reply:
Possibly, we can leave it for a future iteration (and maybe upstream it to the grpc crate if we like it).